feat: Added Support for Fonts

Vishveshwara opened the pull request:
Fixes #239 
Fixes #1134
<!-- Add issue number here. This will automatically closes the issue. If you do not solve the issue entirely, please change the message to e.g. "First steps for issues #IssueNumber" -->

## Changes 
- Created a Drop Down Menu with 7 google Fonts ('Roboto','Open Sans','Lato', 'Poppins','Montserrat','Orbitron', 'Lexend')
- Optimized for characters exceeding the 11x8 grid by combining two characters into an 11x16 layout.
- Dynamically adjusted height to ensure that characters with descenders, such as 'p' and 'q', are fully visible on the screen.
- Enabled effortless font expansion by simply adding new entries to the FontProvider.
- Guaranteed full compatibility with existing emojis while preserving the default font setting for a consistent user experience.

## Screenshots / Recordings  
<!-- Add screen shots/screen recordings of the layout where you made changes or a `*.gif` containing a demonstration. Fill "> N/A" if the change is not a UI fix. -->


https://github.com/user-attachments/assets/59240dea-53e3-46a6-b267-309a8d4de304


**Checklist**: <!-- Please tick following check boxes with `[x]` if the respective task is completed -->
- [x] **No hard coding**: I have used resources from `constants.dart` without hard coding any value.
- [x] **No end of file edits**: No modifications done at end of resource files.
- [x] **Code reformatting**: I have reformatted code and fixed indentation in every file included in this pull request.
- [x] **Code analyzation**: My code passes analyzations run in _flutter analyze_ and tests run in _flutter test_.

## Summary by Sourcery

Add support for custom fonts in the badge display, allowing users to select from a predefined list of Google Fonts and dynamically render text with improved character rendering

New Features:
- Implemented a dropdown menu to select from 7 different Google Fonts
- Added dynamic font rendering that supports characters with descenders
- Created a flexible font system that can easily expand with new font options

Enhancements:
- Optimized text rendering for LED matrix display
- Implemented intelligent character scaling for wide characters
- Added dynamic height adjustment for characters with descenders

Vishveshwara committed:
feat: Added Support for Fonts

Vishveshwara commented:
@samruddhi-Rahegaonkar The issue with spacing of characters is due to the width of characters like "w","h","m" which does not fit in 11x8 grid which is the size we are using for the default font. so for now , I have written a code in which a wide character can occupy 11x16 grid and hence the character does not get cut off but the character occupies double the space.

https://github.com/user-attachments/assets/b36bf056-7264-405f-8b77-5f4e192a88d8

the Solutions for this issue: 
1) Using a smaller font (But causes poor visibility of text. Currently I am using font size 16, font size 14 seems to be fine but still has the spacing issue and anything lower is a problem)
2) use 11x16 grid for all the characters (but the space between characters is very big, can be good for very wide fonts).
3) using a variable grid dimension based on the character width (Best possible solution, but would need a different pipeline from how we are processing the text now)


I will try to work on this, @Jhalakupadhyay can you guide me on which possible solution would be the best

samruddhi-Rahegaonkar commented:
@Vishveshwara That’s a good temporary fix, but it introduces an inconsistency where some characters take double the space, breaking uniform alignment.

samruddhi-Rahegaonkar commented:
@Vishveshwara you might try 
Scale down wide characters to fit within 11x8 without cutting off.

Jhalakupadhyay commented:
@Vishveshwara nice going but there are few changes needed.
1. as suggested by @samruddhi-Rahegaonkar the font should also change in the textfield.
2. Try to maintain a cache for the characters already render such that we don't need to process the characters whose hex is already generated once this will increase the speed of our solution 
3. Let's try scalling down the characters to fit into 11x8 grid and see how it looks I think it will be good option.

I will look closely to PR once and then can tell you better, but we can go for the above changes first.

Vishveshwara committed:
feat: optimized font width calculation and character caching

Vishveshwara commented:
https://github.com/user-attachments/assets/dc1c95ae-23b5-41df-ac77-5a65499b2c67

Challenges
- scaling down the characters to fit into 11x8 grid , makes the characters unreadable so I had to go with the variable column solution

Changes
1. Added Consistent spacing between characters by using variable columns for each character( for example 'i' uses 3 columns and 'W' uses 12 columns.
2. Added Caching for all the fonts currently, and added a feature to generate cache in settings screen for newer fonts.
3. Cache is currently stored as a 2d Binary Bool array.
4. Font changes in the text field.

adityastic commented:
Why do we have 13k loc changed?

adityastic reviewed (changes requested):
I don't think we need 13k lines to have font support in our app.

Vishveshwara commented:
> I don't think we need 13k lines to have font support in our app.

> @Vishveshwara nice going but there are few changes needed.
> 
> 1. as suggested by @samruddhi-Rahegaonkar the font should also change in the textfield.
> 2. Try to maintain a cache for the characters already render such that we don't need to process the characters whose hex is already generated once this will increase the speed of our solution
> 3. Let's try scalling down the characters to fit into 11x8 grid and see how it looks I think it will be good option.
> 
> I will look closely to PR once and then can tell you better, but we can go for the above changes first.

@adityastic
The 13k lines are a result of caching font characters as 2d bool matrices in lib/badgemagic_module/utils/font_cache.dart. As suggested by Jhalak, I implemented a caching mechanism to avoid regenerating character matrices repeatedly.

adityastic commented:
> > I don't think we need 13k lines to have font support in our app.
> 
> > @Vishveshwara nice going but there are few changes needed.
> > 
> > 1. as suggested by @samruddhi-Rahegaonkar the font should also change in the textfield.
> > 2. Try to maintain a cache for the characters already render such that we don't need to process the characters whose hex is already generated once this will increase the speed of our solution
> > 3. Let's try scalling down the characters to fit into 11x8 grid and see how it looks I think it will be good option.
> > 
> > I will look closely to PR once and then can tell you better, but we can go for the above changes first.
> 
> @adityastic
> The 13k lines are a result of caching font characters as 2d bool matrices in **_lib/badgemagic_module/utils/font_cache.dart_**. As suggested by Jhalak, I implemented a caching mechanism to avoid regenerating character matrices repeatedly.

I don't think he meant caching it directly in the repository 😅

Vishveshwara commented:
> I don't think he meant caching it directly in the repository 😅
@adityastic 
I'm sorry for the confusion. I noticed that the default font was being cached in "lib/badgemagic_module/utils/data_to_bytearray_converter.dart", and since the characters were already being cached there, I assumed it was acceptable to cache them directly. However, I understand that this might not be the ideal approach. The code can work without caching as well, so I would appreciate your guidance on where the best place to cache these characters would be.

adityastic commented:
> > I don't think he meant caching it directly in the repository 😅
> 
> @adityastic 
> 
> I'm sorry for the confusion. I noticed that the default font was being cached in "lib/badgemagic_module/utils/data_to_bytearray_converter.dart", and so I assumed it was acceptable to cache them directly. However, I understand that this might not be the ideal approach. The code can work without caching as well, so I would appreciate your guidance on where the best place to cache these characters would be.

We should cache every character in memory while conversation is being performed and reuse the character or the subset if we see repeated entries to improve execution time. That's all to it I guess. Let me know if you have any other thoughts.

Vishveshwara committed:
Updated caching approach

Vishveshwara committed:
Removed font_cache.dart

Vishveshwara committed:
Updated comments in convertBitmapToLEDHex

Vishveshwara commented:
@adityastic can you please review this pr?

Vishveshwara committed:
Merge branch 'flutter_app' into fonts

Vishveshwara commented:
@nope3472 i have read your scrum update, as you are working on the font change feature , you can check this PR out and mention your comments.

Vishveshwara committed:
Merge branch 'development' into fonts

Vishveshwara committed:
fix common build

Vishveshwara committed:
Merge branch 'fonts' of https://github.com/Vishveshwara/badgemagic-app into fonts

Vishveshwara committed:
Fixed common build

samruddhi-Rahegaonkar commented:
Hello, @Vishveshwara Have you tested this implementation on badge ?

samruddhi-Rahegaonkar commented:
> > Hello, @Vishveshwara Have you tested this implementation on badge ?
> 
> Not yet , I will test it today and update you

yup! It will be great to review further!

Vishveshwara committed:
Fixed Artifacts Issue

samruddhi-Rahegaonkar commented on lib/bademagic_module/utils/converters.dart:
renderTextToMatrix() calls toImage() and byte analysis for every character heavy for long strings to optimize it you can Render entire string at once, Split based on change in descender requirement, so try to optimize code as much as you can.

Vishveshwara replied:
I have optimized renderTextToMatrix() Converting the pixel data to boolean matrix code.

For each character toImage() is called just once , and the cache will be saved for each character, we can't cache the entire string. The current method is the best way to do it , Please go through the code line by line and understand it.

and the image size is very less 11x16 at max , this is very fast to compute.

samruddhi-Rahegaonkar commented on lib/bademagic_module/utils/converters.dart:
Try to optimize _processCustomFontMessage and _processDefaultFont into a single function.

Vishveshwara replied:
This shouldn't matter as if a font is not selected , _processDefaultFont is called which uses the method we originally have to process the text(using `converter.charCodes[char]`) just segmenting is added to handle bitmaps and texts.

Vishveshwara committed:
Optimized renderTextToMatrix' For loops

samruddhi-Rahegaonkar reviewed (approved):
LGTM! Nice work.

hpdang commented:
@nope3472 please review this

Vishveshwara committed:
Merge branch 'development' into fonts

Jhalakupadhyay commented:
@Vishveshwara Looks nice have you tested it thoroughly with all the fonts and all teh characters ? If yes I think we can go ahead with this PR.

Vishveshwara committed:
Merge branch 'development' into fonts

Vishveshwara committed:
Applied dart formatting

Vishveshwara committed:
Merge branch 'development' into fonts

samruddhi-Rahegaonkar commented:
Hey, @Vishveshwara Will you please resolve conflicts once.

Vishveshwara committed:
Merge branch 'development' into fonts

Vishveshwara committed:
Applied Dart formatting to homescreen.dart
